packages/workspace/src/tasks-runner/cache.ts

```diff
@@ -0,0 +1,120 @@
+import { appRootPath } from '../utils/app-root';
+import { ProjectGraph } from '../core/project-graph';
+import { NxJson } from '../core/shared-interfaces';
+import { Task } from './tasks-runner';
+import {
+  existsSync,
+  mkdirSync,
+  readFileSync,
+  rmdirSync,
+  writeFileSync
+} from 'fs';
+import { join } from 'path';
+import { Hasher } from './hasher';
+import * as fsExtra from 'fs-extra';
+import { DefaultTasksRunnerOptions } from './tasks-runner-v2';
+
+export type CachedResult = { terminalOutput: string; outputsPath: string };
+export type TaskWithCachedResult = { task: Task; cachedResult: CachedResult };
+
+export class Cache {
+  root = appRootPath;
+  cachePath = this.createCacheDir();
+  hasher = new Hasher(this.projectGraph, this.nxJson);
+
+  constructor(
+    private readonly projectGraph: ProjectGraph,
+    private readonly nxJson: NxJson,
+    private readonly options: DefaultTasksRunnerOptions
+  ) {}
+
+  async get(task: Task): Promise<CachedResult> {
+    if (!this.cacheable(task)) return Promise.resolve(null);
+
+    const res = this.getFromLocalDir(task);
+    if (!res && this.options.remoteCache) {
+      await this.options.remoteCache.retrieve(
+        this.hasher.hash(task),
+        this.cachePath
+      );
+      return this.getFromLocalDir(task);
+    } else {
+      return res;
+    }
+  }
+
+  put(task: Task, terminalOutput: string, folders: string[]) {
+    if (!this.cacheable(task)) return;
+    const td = this.taskDir(task);
+    mkdirSync(td);
+    writeFileSync(join(td, 'terminalOutput'), terminalOutput);
+
+    mkdirSync(join(td, 'outputs'));
+    folders.forEach(f => {
+      const srcDir = join(this.root, f);
+      if (existsSync(srcDir)) {
+        const cachedDir = join(td, 'outputs', f);
+        mkdirSync(cachedDir, { recursive: true });
+        fsExtra.copySync(srcDir, cachedDir);
+      }
+    });
+
+    if (this.options.remoteCache) {
+      this.options.remoteCache.store(this.hasher.hash(task), this.cachePath);
+    }
+  }
+
+  copyFilesFromCache(cachedResult: CachedResult, outputs: string[]) {
+    outputs.forEach(f => {
+      const cachedDir = join(cachedResult.outputsPath, f);
+      if (existsSync(cachedDir)) {
+        const srcDir = join(this.root, f);
+        try {
+          rmdirSync(srcDir);
```

Comment:
Should this be `rmdirSync(srcDir, { recursive: true })` in case it doesn't exist or is empty?

Should the following two lines (75-76) be in the try block as well since the rm failing may mean they will fail as well?

Reply:
rmdir doesn't take `recursive: true`. It removes the dir with all its contents.

The idea is that `rmdirSync` should only fail if the dir isn't present. In this case, `mkdirSync` won't fail.  I'm going to change it to remove the try-catch.

Reply:
Change it to: 

```
        const srcDir = join(this.root, f);
        if (existsSync(srcDir)) {
          rmdirSync(srcDir);
        }
        mkdirSync(srcDir, { recursive: true });
        fsExtra.copySync(cachedDir, srcDir);
```